network/phonebook.go

```diff
@@ -35,17 +35,40 @@ type Phonebook interface {
 
 	// UpdateRetryAfter updates the retry-after field for the entries matching the given address
 	UpdateRetryAfter(addr string, retryAfter time.Time)
+
+	// waitAndAddConnectionTime will wait to prevent exceeding connectionsRateLimitingCount.
+	// Then it will register the next connection time.
+	WaitAndAddConnectionTime(addr string,
+		connectionsRateLimitingCount uint, connectionsRateLimitingWindowSeconds uint)
 }
 
 type phonebookData struct {
-	retryAfter time.Time
+	retryAfter            time.Time
+	recentConnectionTimes []time.Time
 }
 
 type phonebookEntries map[string]phonebookData
```

Comment:
I think that you'd like to change the phonebookEntries to be a struct, so that it could hold the connectionsRateLimitingCount  and connectionsRateLimitingWindowSeconds.

That would allow you to pass these during the construction rather than over and over.